### DISCUSSION TITLE
decreased reliability when using ridge regression

### DISCUSSION THREAD

**Discussion by zs-nemecz** [cvnlab/GLMsingle#110] [Q&A] [answered]
Hello! 

Thanks for creating this toolbox, great work.
 
I tried running GLMsingle on our data using HRF fitting, GLMdenoise, and ridge regression, as described in example1 (python version). What I first noticed is that the R2 and FRAC values look very ‘noisy’. I attributed this to the fact that we have substantially less data than in example1 (2 runs with 250 volumes, TR of 1.84s). I get a very similar pattern of results when I upsample the data and the design matrix to a TR of 1s, as recommended in some other threads. 

![kép](https://github.com/cvnlab/GLMsingle/assets/26097191/710550de-0862-4b31-a0eb-96649ff6cfb2)

But what I found even stranger is that the reliability calculated on repeated trials is very low (below or close to 0) with the RR model compared to FITHRF and GLMdenoise. This effect seems to be independent of the ROI I use.

![kép](https://github.com/cvnlab/GLMsingle/assets/26097191/a7787f92-0889-4813-bb1d-fe001eaa9fda)
![kép](https://github.com/cvnlab/GLMsingle/assets/26097191/4d70d4c0-685f-4a00-af28-da1b97331e72)

Am I doing something wrong here? What are some potential reasons why reliability decreases so much in the RR case compared to the other models?

**Comment by kendrickkay** [comment, marked as answer]
Hi. Thanks for the informative description. Based on the first figure, I would say that actually your results look fairly good.  There is a clear signal in the R2 and FRACvalue plots (and this is good). (Some datasets fare much worse.).   If you are specifically curious about the "speckly" nature of the R2 and FRACvalue plots, it is true that this exists. And it is true that the cause of this is likely to be the relatively few number of time points that went into the analysis (the idea being, that the fewer the number of time points the less stable are the measures that one derives from a given dataset). However, it isn't necessarily the case that anything is "wrong"/"buggy", but it is what it is.

I agree that it seems quite strange/surprising that the RR results for the reliability seem to go down.  This, in general, is a tricky area to wade through, and you have to carefully consider different possibilities.

One key observation that does suggest things should be working well is how the FRACvalue plot looks.  According to that plot, there are a number of voxels that are successfully being labeled as "not needing much regularization" (since their fractions are close to 1).  This suggests that the data and setup are fundamentally sound.

It is true, however, that the vast majority of voxels in your volume do not appear to have reliable experiment-related signals. This is quite typical (since most of the voxels are outside the brain or in white matter or in parts of the brain that do not care about your experiment).

For better diagnosis, consider plotting some scatter plots where the x-axis contains e.g. the GLMdenoise reliability values and the y-axis contains the RR reliability values.   The idea here is that we really want to pinpoint exactly where the "good" voxels are.  We would not want to be unduly confused by the vast population of bad voxels.  (How does reliability change for the very best voxels in your dataset?)

That being said, it is indeed strange that the central tendency shifts from around 0.1 down to 0.  How exactly are you computing the reliability?  Are the trials that contribute to this randomized or split by run?  Could there be something funny in terms of temporal correlation / temporal context that somehow inflates irrelevant voxels to have positive reliability?  (For example, if you inspect a volume of your FITHRF and/or FITHRF_GLMDENOISE reliability values, do they seem to be positive even in parts of the brain that should have no signal?)

Of course, one option at your disposal is to simply go with, e.g., the FITHRF_GLMDENOISE beta values.  But, if you are curious to get to the bottom of things, it would be more satisfying to understand what is going on.

**Comment by zs-nemecz** [comment]
Hi, thanks for the detailed response and sorry for taking so long to get back to this. 

Here I plotted the GLMdenoise and RR reliability values as suggested. I used the reliability values from one ROI for plotting, but the relationship is similar in other parts of the brain. 

![kép](https://github.com/cvnlab/GLMsingle/assets/26097191/566d40e4-76f9-4b34-b3b0-8dbaf1193f14)

There is indeed something special about the trials that go into the reliability analysis. These are repeated trials, and the repetitions are pretty close to each other in time, i.e. 14 TRs apart on average, and only a few trials are repeated across runs. So I guess the temporal proximity contributes to the inflated reliability, and it also seems to be the case that the FITHRF_GLMDENOISE reliability values tend to be positive even in voxels where I would not expect a signal. This is less so with the RR model.  
I went back to the autocorrelation and neighboring trials section of the eLife paper, and the results sort of make sense in hindsight. 
Our repeated trials are not "neighboring" per se but many are close enough to be affected by the autocorrelation problem.

**Reply by kendrickkay** [reply]
In your scatter plot, the generally high correlation between the two analysis variants is a good sign that things are sane.

Your idea about the temporal proximity does sound plausible.  Note that things depend on how exactly you are computing reliability (e.g., how exactly do the repetitions get split into different things that you correlate)...   But yes, in general, trials that tend to be close together in time are generally expected to be similar in their signal magnitude (in fMRI data).

Also, as you point out, it is important to consider what is happening in voxels where you dont' expect a signal.  (In other words, just because we devise a metric that we "want" to be high doesn't mean that we want it to be high in all cases... we want it to be high only for genuine reasons!)

If you are still worried, there are other analyses/visualizations that you could try (as I mentioned in the earlier message)
